File: CBTReframe/Services/LLMJSONSanitizer.swift
import Foundation

/// Gemini 等模型偶发输出残缺 JSON、混用中英文键名或夹杂说明文字；在解析前尽量裁出可解析片段。
enum LLMJSONSanitizer {
    static func sanitizeForJSONObject(_ raw: String) -> String {
        var text = raw
            .replacingOccurrences(of: "```json", with: "")
            .replacingOccurrences(of: "```JSON", with: "")
            .replacingOccurrences(of: "```", with: "")
            .trimmingCharacters(in: .whitespacesAndNewlines)

        // 取第一个 { 与最后一个 } 之间的内容（忽略尾部截断时可能的多余字符）
        if let start = text.firstIndex(of: "{"),
           let end = text.lastIndex(of: "}") {
            text = String(text[start...end])
        }

        // 常见：未闭合字符串 — 尝试补全引号与括号（保守）
        text = repairTruncatedJSON(text)

        return text
    }

    /// 若末尾在字符串内被截断，补一个 `"` 并尝试闭合 `}`。
    private static func repairTruncatedJSON(_ s: String) -> String {
        var t = s
        let open = t.filter { $0 == "{" }.count
        let close = t.filter { $0 == "}" }.count
        if open > close {
            // 简单补全：若最后一个非空白字符不是 }，先尝试闭合字符串再补 }
            let trimmed = t.trimmingCharacters(in: .whitespacesAndNewlines)
            if let last = trimmed.last, last != "}" {
                if last != "\"" {
                    t += "\""
                }
                t += String(repeating: "}", count: open - close)
            }
        }
        return t
    }
}

File: shared/src/commonMain/kotlin/com/henryliu/cbtreframe/shared/LLMJSONSanitizer.kt
package com.henryliu.cbtreframe.shared

object LLMJSONSanitizer {
    /**
     * Strip markdown fences, extract the outermost JSON object between the
     * first `{` and last `}`, then repair truncation by counting braces and
     * appending missing quotes / closing braces.
     */
    fun sanitizeForJSONObject(raw: String): String {
        var text = raw
            .replace("```json", "")
            .replace("```JSON", "")
            .replace("```", "")
            .trim()

        val startIdx = text.indexOf('{')
        val endIdx = text.lastIndexOf('}')
        if (startIdx != -1 && endIdx != -1 && endIdx > startIdx) {
            text = text.substring(startIdx, endIdx + 1)
        }

        text = repairTruncatedJSON(text)
        return text
    }

    // ── Private helpers ────────────────────────────────────────────────

    /**
     * Walk the JSON string character-by-character, tracking whether we are
     * inside a string literal (respecting backslash escapes).  Count open
     * `{` / `}` to compute the required closing-brace deficit.
     *
     * If the last meaningful token is an incomplete string (e.g. `"key":`
     * or `"value`), append a closing quote before the missing braces.
     */
    private fun repairTruncatedJSON(s: String): String {
        var braceDepth = 0
        var inString = false
        var escape = false
        var lastNonWhitespace: Char? = null

        for (ch in s) {
            if (escape) {
                escape = false
                continue
            }
            when {
                ch == '\\' && inString -> escape = true
                ch == '"' -> inString = !inString
                ch == '{' && !inString -> braceDepth++
                ch == '}' && !inString -> braceDepth--
            }
            if (!ch.isWhitespace()) lastNonWhitespace = ch
        }

        if (braceDepth <= 0) return s

        var result = s

        // Append a closing quote if we ended inside an unclosed string.
        if (inString && lastNonWhitespace != '"') {
            result += "\""
        }

        // Close remaining open braces.
        result += "}".repeat(braceDepth)

        return result
    }
}
